packages/skygear-react-native/ios/SGSkygearReactNative.m

```diff
@@ -0,0 +1,353 @@
+@interface SGSkygearReactNative()
+// We must have strong reference to the session otherwise it is closed immediately when
+// it goes out of scope.
+@property (nonatomic, strong) SFAuthenticationSession *sfSession API_AVAILABLE(ios(11));
+@end
+#endif
+
+#if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= 12000)
+@interface SGSkygearReactNative() <ASWebAuthenticationPresentationContextProviding>
+// We must have strong reference to the session otherwise it is closed immediately when
+// it goes out of scope.
+@property (nonatomic, strong) ASWebAuthenticationSession *asSession API_AVAILABLE(ios(12));
+@end
+#endif
+
+#if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= 13000)
+@interface SGSkygearReactNative() <ASAuthorizationControllerPresentationContextProviding, ASAuthorizationControllerDelegate>
+@property (nonatomic, strong) ASAuthorizationAppleIDProvider *provider API_AVAILABLE(ios(13));
+@property (nonatomic, strong) ASAuthorizationController *controller API_AVAILABLE(ios(13));
+@end
+#endif
+
+@implementation SGSkygearReactNative
+
+RCT_EXPORT_MODULE()
+
++ (BOOL)requiresMainQueueSetup
+{
+  return YES;
+}
+
+- (instancetype)init
+{
+    if ((self = [super init])) {
+        [[NSNotificationCenter defaultCenter] addObserver:self
+                                                 selector:@selector(handleOpenURLNotification:)
+                                                     name:kOpenURLNotification
+                                                   object:nil];
+    }
+    return self;
+}
+
+- (dispatch_queue_t)methodQueue
+{
+  return dispatch_get_main_queue();
+}
+
++ (BOOL)application:(UIApplication *)app
+            openURL:(NSURL *)URL
+            options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
+{
+  postNotificationWithURL(URL, self);
+  return YES;
+}
+
++ (BOOL)application:(UIApplication *)application
+            openURL:(NSURL *)URL
+  sourceApplication:(NSString *)sourceApplication
+         annotation:(id)annotation
+{
+  postNotificationWithURL(URL, self);
+  return YES;
+}
+
++ (BOOL)application:(UIApplication *)application
+continueUserActivity:(NSUserActivity *)userActivity
+  restorationHandler:
+    #if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= 12000)
+        (nonnull void (^)(NSArray<id<UIUserActivityRestoring>> *_Nullable))restorationHandler {
+    #else
+        (nonnull void (^)(NSArray *_Nullable))restorationHandler {
+    #endif
+  if ([userActivity.activityType isEqualToString:NSUserActivityTypeBrowsingWeb]) {
+    postNotificationWithURL(userActivity.webpageURL, self);
+  }
+  return YES;
+}
+
+RCT_EXPORT_METHOD(openURL:(NSURL *)url
+                   scheme:(NSString *)scheme
+                  resolve:(RCTPromiseResolveBlock)resolve
+                   reject:(RCTPromiseRejectBlock)reject)
+{
+    self.openURLResolve = resolve;
```

Comment:
I suppose this function does not support reentrancy on iOS? We should reject the previous promise with some error, rather than leaving it hanging.

Reply:
Yes, you're right. It does not support reentrancy. I do not think there is a valid use case to login concurrently.